File: posbillingwala-flutter/lib/features/payment_display/domain/display_connection_manager.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:pos_billingwala_v2/core/logging/app_logger.dart';
import 'package:pos_billingwala_v2/core/utils/app_platform.dart';
import 'package:pos_billingwala_v2/features/payment_display/data/payment_display_server.dart';
import 'package:pos_billingwala_v2/features/payment_display/domain/payment_display_models.dart';
import 'package:pos_billingwala_v2/features/payment_display/domain/payment_display_settings.dart';

class PaymentDisplayUiState {
  const PaymentDisplayUiState({
    this.status = PaymentDisplayConnectionStatus.stopped,
    this.localUrl = '',
    this.pairingUrl = '',
    this.connectedClients = 0,
    this.autoDisplayNewBills = true,
    this.qrDurationSeconds = PaymentDisplaySettings.defaultQrDurationSeconds,
    this.activeBill,
    this.errorMessage,
    this.serverRunning = false,
  });

  final PaymentDisplayConnectionStatus status;
  final String localUrl;
  final String pairingUrl;
  final int connectedClients;
  final bool autoDisplayNewBills;
  final int qrDurationSeconds;
  final PaymentDisplayBillPayload? activeBill;
  final String? errorMessage;
  final bool serverRunning;

  bool get isConnected =>
      status == PaymentDisplayConnectionStatus.connected &&
      connectedClients > 0;

  PaymentDisplayUiState copyWith({
    PaymentDisplayConnectionStatus? status,
    String? localUrl,
    String? pairingUrl,
    int? connectedClients,
    bool? autoDisplayNewBills,
    int? qrDurationSeconds,
    PaymentDisplayBillPayload? activeBill,
    String? errorMessage,
    bool? serverRunning,
    bool clearError = false,
    bool clearBill = false,
  }) {
    return PaymentDisplayUiState(
      status: status ?? this.status,
      localUrl: localUrl ?? this.localUrl,
      pairingUrl: pairingUrl ?? this.pairingUrl,
      connectedClients: connectedClients ?? this.connectedClients,
      autoDisplayNewBills: autoDisplayNewBills ?? this.autoDisplayNewBills,
      qrDurationSeconds: qrDurationSeconds ?? this.qrDurationSeconds,
      activeBill: clearBill ? null : (activeBill ?? this.activeBill),
      errorMessage: clearError ? null : (errorMessage ?? this.errorMessage),
      serverRunning: serverRunning ?? this.serverRunning,
    );
  }
}

/* Owns local server lifecycle, pairing, QR expiry (authoritative), reconnect. */
class DisplayConnectionManager extends Notifier<PaymentDisplayUiState> {
  final _settingsStore = PaymentDisplaySettingsStore();
  PaymentDisplayServer? _server;
  Timer? _expiryTimer;
  Timer? _ipWatchTimer;
  PaymentDisplaySettings _settings = const PaymentDisplaySettings();

  @override
  PaymentDisplayUiState build() {
    ref.onDispose(() {
      _expiryTimer?.cancel();
      _ipWatchTimer?.cancel();
      unawaited(_server?.stop());
    });
    Future.microtask(_bootstrap);
    return const PaymentDisplayUiState();
  }

  Future<void> _bootstrap() async {
    if (kIsWeb || AppPlatform.isWeb) return;
    _settings = await _settingsStore.load();
    PaymentDisplayBillPayload? restored;
    final raw = _settings.lastBillJson.trim();
    if (raw.isNotEmpty) {
      try {
        final map = jsonDecode(raw);
        if (map is Map<String, dynamic>) {
          final bill = PaymentDisplayBillPayload.fromJson(map);
          if (!bill.isExpired) {
            restored = bill;
          } else {
            await _settingsStore.clearLastBill();
          }
        }
      } catch (_) {
        await _settingsStore.clearLastBill();
      }
    }
    state = state.copyWith(
      autoDisplayNewBills: _settings.autoDisplayNewBills,
      qrDurationSeconds: _settings.qrDurationSeconds,
      localUrl: _settings.lastLocalUrl,
      activeBill: restored,
      clearBill: restored == null,
    );
    if (restored != null) {
      _scheduleExpiry(restored.expiresAt);
    }
  }

  PaymentDisplayServer get _ensureServer {
    return _server ??= PaymentDisplayServer(
      onClientCountChanged: (count) {
        final nextStatus = !_server!.isRunning
            ? PaymentDisplayConnectionStatus.stopped
            : count > 0
            ? PaymentDisplayConnectionStatus.connected
            : PaymentDisplayConnectionStatus.waitingForPair;
        state = state.copyWith(
          connectedClients: count,
          status: nextStatus,
          pairingUrl: _server?.pairingUrl ?? state.pairingUrl,
          localUrl: _server?.displayBaseUrl ?? state.localUrl,
        );
      },
      onError: (msg) {
        state = state.copyWith(
          status: PaymentDisplayConnectionStatus.error,
          errorMessage: msg,
        );
      },
    );
  }

  Future<bool> startServer({bool rotateToken = true}) async {
    if (kIsWeb || AppPlatform.isWeb) {
      state = state.copyWith(
        status: PaymentDisplayConnectionStatus.error,
        errorMessage: 'Payment display requires the Android POS app.',
      );
      return false;
    }
    state = state.copyWith(
      status: PaymentDisplayConnectionStatus.starting,
      clearError: true,
    );
    final server = _ensureServer;
    final ok = await server.start();
    if (!ok) {
      state = state.copyWith(
        status: PaymentDisplayConnectionStatus.error,
        serverRunning: false,
        errorMessage: state.errorMessage ?? 'Could not start display server.',
      );
      return false;
    }
    if (rotateToken) server.rotatePairingToken();
    if (state.activeBill != null) {
      server.setActiveBill(state.activeBill);
    }
    _startIpWatch();
    final url = server.displayBaseUrl ?? '';
    final pairing = server.pairingUrl ?? '';
    _settings = _settings.copyWith(lastLocalUrl: url);
    await _settingsStore.save(_settings);
    state = state.copyWith(
      status: server.connectedClientCount > 0
          ? PaymentDisplayConnectionStatus.connected
          : PaymentDisplayConnectionStatus.waitingForPair,
      serverRunning: true,
      localUrl: url,
      pairingUrl: pairing,
      connectedClients: server.connectedClientCount,
      clearError: true,
    );
    return true;
  }

  Future<void> stopServer() async {
    _ipWatchTimer?.cancel();
    _expiryTimer?.cancel();
    await _server?.stop();
    state = state.copyWith(
      status: PaymentDisplayConnectionStatus.stopped,
      serverRunning: false,
      connectedClients: 0,
      pairingUrl: '',
      clearError: true,
    );
  }

  Future<void> showPairingQr() async {
    final running = state.serverRunning
        ? true
        : await startServer(rotateToken: true);
    if (!running) return;
    _server?.rotatePairingToken();
    state = state.copyWith(
      pairingUrl: _server?.pairingUrl ?? '',
      localUrl: _server?.displayBaseUrl ?? state.localUrl,
      status: state.connectedClients > 0
          ? PaymentDisplayConnectionStatus.connected
          : PaymentDisplayConnectionStatus.waitingForPair,
    );
    AppLogger.info('PAYMENT_DISPLAY_PAIRING_STARTED');
  }

  Future<void> reconnect() async {
    await _server?.refreshNetworkIdentity();
    if (!state.serverRunning) {
      await startServer(rotateToken: false);
      return;
    }
    state = state.copyWith(
      localUrl: _server?.displayBaseUrl ?? state.localUrl,
      pairingUrl: _server?.pairingUrl ?? state.pairingUrl,
      clearError: true,
    );
    AppLogger.info('PAYMENT_DISPLAY_RECONNECT');
  }

  Future<void> disconnectDisplays() async {
    await _server?.disconnectAllClients();
    state = state.copyWith(
      connectedClients: 0,
      status: state.serverRunning
          ? PaymentDisplayConnectionStatus.waitingForPair
          : PaymentDisplayConnectionStatus.stopped,
    );
  }

  Future<void> setAutoDisplayNewBills(bool value) async {
    _settings = _settings.copyWith(autoDisplayNewBills: value);
    await _settingsStore.save(_settings);
    state = state.copyWith(autoDisplayNewBills: value);
  }

  Future<void> setQrDurationSeconds(int seconds) async {
    final clamped = seconds.clamp(
      PaymentDisplaySettings.minQrDurationSeconds,
      PaymentDisplaySettings.maxQrDurationSeconds,
    );
    _settings = _settings.copyWith(qrDurationSeconds: clamped);
    await _settingsStore.save(_settings);
    state = state.copyWith(qrDurationSeconds: clamped);
  }

  Duration get qrDuration => _settings.qrDuration;

  bool get autoDisplayEnabled => _settings.autoDisplayNewBills;

  Future<ShowPaymentDisplayResult> publishBill(
    PaymentDisplayBillPayload bill,
  ) async {
    if (kIsWeb || AppPlatform.isWeb) {
      return const ShowPaymentDisplayResult(
        code: ShowPaymentDisplayResultCode.notSupported,
        message: 'Payment display requires the Android POS app.',
      );
    }
    if (!state.serverRunning || _server == null || !_server!.isRunning) {
      return const ShowPaymentDisplayResult(
        code: ShowPaymentDisplayResultCode.displayNotConnected,
        message: 'Payment display is not connected.',
      );
    }
    if (state.connectedClients < 1) {
      return const ShowPaymentDisplayResult(
        code: ShowPaymentDisplayResultCode.displayNotConnected,
        message: 'Payment display is not connected.',
      );
    }

    _expiryTimer?.cancel();
    _server!.setActiveBill(bill);
    await _server!.broadcastBillUpdated(bill);
    _scheduleExpiry(bill.expiresAt);
    state = state.copyWith(activeBill: bill, clearError: true);
    _settings = _settings.copyWith(lastBillJson: jsonEncode(bill.toJson()));
    await _settingsStore.save(_settings);
    return ShowPaymentDisplayResult(
      code: ShowPaymentDisplayResultCode.success,
      payload: bill,
    );
  }

  Future<void> clearBill({String reason = 'expired'}) async {
    _expiryTimer?.cancel();
    if (_server != null && _server!.isRunning) {
      await _server!.broadcastBillCleared(reason: reason);
    }
    state = state.copyWith(clearBill: true);
    await _settingsStore.clearLastBill();
    _settings = _settings.copyWith(clearLastBill: true);
  }

  void _scheduleExpiry(DateTime expiresAt) {
    _expiryTimer?.cancel();
    final wait = expiresAt.difference(DateTime.now());
    if (wait <= Duration.zero) {
      unawaited(clearBill(reason: 'expired'));
      return;
    }
    _expiryTimer = Timer(wait, () {
      unawaited(clearBill(reason: 'expired'));
    });
  }

  void _startIpWatch() {
    _ipWatchTimer?.cancel();
    _ipWatchTimer = Timer.periodic(const Duration(seconds: 20), (_) async {
      if (_server == null || !_server!.isRunning) return;
      await _server!.refreshNetworkIdentity();
      final url = _server!.displayBaseUrl ?? state.localUrl;
      final pairing = _server!.pairingUrl ?? state.pairingUrl;
      if (url != state.localUrl || pairing != state.pairingUrl) {
        state = state.copyWith(localUrl: url, pairingUrl: pairing);
        _settings = _settings.copyWith(lastLocalUrl: url);
        await _settingsStore.save(_settings);
      }
    });
  }
}

final displayConnectionManagerProvider =
    NotifierProvider<DisplayConnectionManager, PaymentDisplayUiState>(
      DisplayConnectionManager.new,
    );

File: WithTable/app/src/main/java/com/pos_billingwala/PaymentDisplay/DisplayConnectionManager.java
package com.pos_billingwala.PaymentDisplay;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import org.json.JSONObject;

import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Singleton: server lifecycle, pairing, authoritative 5-minute QR expiry.
 */
public final class DisplayConnectionManager {

    private static final String TAG = "PaymentDisplay";

    public enum Status {
        STOPPED, STARTING, WAITING_FOR_PAIR, CONNECTED, ERROR
    }

    public interface Listener {
        void onDisplayStateChanged();
    }

    private static DisplayConnectionManager instance;

    public static synchronized DisplayConnectionManager get(Context context) {
        if (instance == null) {
            instance = new DisplayConnectionManager(context.getApplicationContext());
        }
        return instance;
    }

    private final Context appContext;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private final CopyOnWriteArrayList<Listener> listeners = new CopyOnWriteArrayList<>();

    private PaymentDisplayServer server;
    private Status status = Status.STOPPED;
    private String errorMessage;
    private int connectedClients;
    private String pairingUrl = "";
    private String localUrl = "";
    private PaymentDisplayBillPayload activeBill;
    private final Runnable expiryRunnable = this::onExpiry;
    private final Runnable ipWatchRunnable = this::onIpWatch;

    private DisplayConnectionManager(Context context) {
        this.appContext = context.getApplicationContext();
        restoreBillIfValid();
    }

    public void addListener(Listener listener) {
        if (listener != null) {
            listeners.add(listener);
        }
    }

    public void removeListener(Listener listener) {
        listeners.remove(listener);
    }

    public Status getStatus() {
        return status;
    }

    public String getErrorMessage() {
        return errorMessage;
    }

    public int getConnectedClients() {
        return connectedClients;
    }

    public String getPairingUrl() {
        return pairingUrl;
    }

    public String getLocalUrl() {
        return localUrl;
    }

    public PaymentDisplayBillPayload getActiveBill() {
        return activeBill;
    }

    public boolean isServerRunning() {
        return server != null && server.isRunning();
    }

    public boolean isConnected() {
        return isServerRunning() && connectedClients > 0;
    }

    public boolean isAutoDisplayEnabled() {
        return PaymentDisplayPrefs.isAutoDisplayEnabled(appContext);
    }

    public void setAutoDisplayEnabled(boolean enabled) {
        PaymentDisplayPrefs.setAutoDisplayEnabled(appContext, enabled);
        notifyListeners();
    }

    public int getQrDurationSeconds() {
        return PaymentDisplayPrefs.getQrDurationSeconds(appContext);
    }

    public void setQrDurationSeconds(int seconds) {
        PaymentDisplayPrefs.setQrDurationSeconds(appContext, seconds);
        notifyListeners();
    }

    public synchronized boolean startServer(boolean rotateToken) {
        status = Status.STARTING;
        errorMessage = null;
        notifyListeners();
        if (server == null) {
            server = new PaymentDisplayServer(appContext, count -> mainHandler.post(() -> {
                connectedClients = count;
                status = !isServerRunning()
                        ? Status.STOPPED
                        : (count > 0 ? Status.CONNECTED : Status.WAITING_FOR_PAIR);
                pairingUrl = server != null && server.getPairingUrl() != null ? server.getPairingUrl() : pairingUrl;
                localUrl = server != null && server.getDisplayBaseUrl() != null ? server.getDisplayBaseUrl() : localUrl;
                notifyListeners();
            }));
        }
        boolean ok = server.startServer();
        if (!ok) {
            status = Status.ERROR;
            errorMessage = "Could not start display server. Enable hotspot/Wi-Fi.";
            notifyListeners();
            return false;
        }
        if (rotateToken) {
            server.rotatePairingToken();
        }
        if (activeBill != null && !activeBill.isExpired()) {
            server.setActiveBill(activeBill);
        }
        pairingUrl = server.getPairingUrl() != null ? server.getPairingUrl() : "";
        localUrl = server.getDisplayBaseUrl() != null ? server.getDisplayBaseUrl() : "";
        PaymentDisplayPrefs.setLastLocalUrl(appContext, localUrl);
        connectedClients = server.getConnectedClientCount();
        status = connectedClients > 0 ? Status.CONNECTED : Status.WAITING_FOR_PAIR;
        mainHandler.removeCallbacks(ipWatchRunnable);
        mainHandler.postDelayed(ipWatchRunnable, 20_000);
        notifyListeners();
        return true;
    }

    public synchronized void stopServer() {
        mainHandler.removeCallbacks(expiryRunnable);
        mainHandler.removeCallbacks(ipWatchRunnable);
        if (server != null) {
            server.stopServer();
        }
        status = Status.STOPPED;
        connectedClients = 0;
        pairingUrl = "";
        notifyListeners();
    }

    public synchronized void showPairingQr() {
        if (!isServerRunning()) {
            startServer(true);
        } else if (server != null) {
            server.rotatePairingToken();
            pairingUrl = server.getPairingUrl() != null ? server.getPairingUrl() : "";
            localUrl = server.getDisplayBaseUrl() != null ? server.getDisplayBaseUrl() : localUrl;
            status = connectedClients > 0 ? Status.CONNECTED : Status.WAITING_FOR_PAIR;
            notifyListeners();
        }
    }

    public synchronized void reconnect() {
        if (server != null) {
            server.refreshIp();
        }
        if (!isServerRunning()) {
            startServer(false);
            return;
        }
        pairingUrl = server.getPairingUrl() != null ? server.getPairingUrl() : pairingUrl;
        localUrl = server.getDisplayBaseUrl() != null ? server.getDisplayBaseUrl() : localUrl;
        PaymentDisplayPrefs.setLastLocalUrl(appContext, localUrl);
        Log.i(TAG, "PAYMENT_DISPLAY_RECONNECT");
        notifyListeners();
    }

    public synchronized void disconnectDisplays() {
        if (server != null) {
            server.disconnectAllClients();
        }
        connectedClients = 0;
        status = isServerRunning() ? Status.WAITING_FOR_PAIR : Status.STOPPED;
        notifyListeners();
    }

    public synchronized ShowResult publishBill(PaymentDisplayBillPayload bill) {
        if (!isServerRunning() || server == null) {
            return ShowResult.notConnected();
        }
        if (connectedClients < 1) {
            return ShowResult.notConnected();
        }
        mainHandler.removeCallbacks(expiryRunnable);
        server.broadcastBillUpdated(bill);
        activeBill = bill;
        PaymentDisplayPrefs.setLastBillJson(appContext, bill.toPersistJson().toString());
        long delay = Math.max(0L, bill.expiresAtMs - System.currentTimeMillis());
        mainHandler.postDelayed(expiryRunnable, delay);
        notifyListeners();
        return ShowResult.success(bill);
    }

    public synchronized void clearBill(String reason) {
        mainHandler.removeCallbacks(expiryRunnable);
        if (server != null && server.isRunning()) {
            server.broadcastBillCleared(reason);
        }
        activeBill = null;
        PaymentDisplayPrefs.clearLastBill(appContext);
        notifyListeners();
    }

    private void onExpiry() {
        clearBill("expired");
    }

    private void onIpWatch() {
        if (!isServerRunning()) {
            return;
        }
        reconnect();
        mainHandler.postDelayed(ipWatchRunnable, 20_000);
    }

    private void restoreBillIfValid() {
        try {
            String raw = PaymentDisplayPrefs.getLastBillJson(appContext);
            if (raw == null || raw.trim().isEmpty()) {
                return;
            }
            PaymentDisplayBillPayload bill = PaymentDisplayBillPayload.fromJson(new JSONObject(raw));
            if (bill != null && !bill.isExpired()) {
                activeBill = bill;
                long delay = bill.remainingMs();
                mainHandler.postDelayed(expiryRunnable, delay);
            } else {
                PaymentDisplayPrefs.clearLastBill(appContext);
            }
        } catch (Exception e) {
            PaymentDisplayPrefs.clearLastBill(appContext);
        }
        localUrl = PaymentDisplayPrefs.getLastLocalUrl(appContext);
    }

    private void notifyListeners() {
        for (Listener l : listeners) {
            try {
                l.onDisplayStateChanged();
            } catch (Exception ignored) {
            }
        }
    }

    public static final class ShowResult {
        public enum Code {
            SUCCESS, NOT_CONNECTED, UPI_NOT_CONFIGURED, INVALID_AMOUNT,
            INVOICE_INVALID, QR_FAILED, ERROR
        }

        public final Code code;
        public final String message;
        public final PaymentDisplayBillPayload payload;

        private ShowResult(Code code, String message, PaymentDisplayBillPayload payload) {
            this.code = code;
            this.message = message;
            this.payload = payload;
        }

        public static ShowResult success(PaymentDisplayBillPayload p) {
            return new ShowResult(Code.SUCCESS, null, p);
        }

        public static ShowResult notConnected() {
            return new ShowResult(Code.NOT_CONNECTED, "Payment display is not connected.", null);
        }

        public static ShowResult fail(Code code, String message) {
            return new ShowResult(code, message, null);
        }

        public boolean isSuccess() {
            return code == Code.SUCCESS;
        }
    }
}
